### PR TITLE
Bug fixes for intelephense and CCLS

### LINKED ISSUE #498: CCLS support / "complete_using_text_edit"
https://github.com/MaskRay/ccls/issues/226

## Currently
```c
client->a
```

I get autocompletion like this: `a : int tcp_client_s`
After selecting that autocompletion the code looks like this:
```c
client->a : int
```

#### OS and language server:
ArchLinux with ccls for C/C++/ObjC (git version installed a5f600a7)
https://github.com/MaskRay/ccls

#### How you installed LSP:
Package Control

#### Client configuration:
```
"ccls":
{
  "command":
  [
    "ccls"
  ],
  "initializationOptions":
  {
    "cacheDirectory": "/tmp/ccls"
  },
  "languages":
  [
    {
      "languageId": "c",
      "scopes":
      [
        "source.c"
      ],
      "syntaxes":
      [
        "Packages/C++/C.sublime-syntax"
      ]
    },
    {
      "languageId": "cpp",
      "scopes":
      [
        "source.c++"
      ],
      "syntaxes":
      [
        "Packages/C++/C++.sublime-syntax"
      ]
    },
    {
      "languageId": "objective-c",
      "scopes":
      [
        "source.objc"
      ],
      "syntaxes":
      [
        "Packages/Objective-C/Objective-C.sublime-syntax"
      ]
    },
    {
      "languageId": "objective-cpp",
      "scopes":
      [
        "source.objc++"
      ],
      "syntaxes":
      [
        "Packages/Objective-C/Objective-C++.sublime-syntax"
      ]
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by rwols**
This pull request supersedes #521 and #520 as it contains the cherry-picked commits from them.

It should resolve #498, #519, #517.

- Fix up textDocument/documentSymbol request for newer protocol versions.
- Fix up textDocument/signatureHelp request for newer protocol versions.
- Prefer textEdit over insertText, also for newer protocol versions.

I also added completion tests for intelephense, clangd, and pyls.

### CODE REVIEW

File: plugin/symbols.py [hunk 1]
```diff
@@ -81,7 +82,13 @@ def handle_response(self, response: 'Optional[List]') -> None:
 
     def on_symbol_selected(self, symbol_index):
         selected_symbol = self.symbols[symbol_index]
-        range = selected_symbol['location']['range']
+        try:
+            range = selected_symbol['location']['range']
+        except KeyError:
```
Comment (tomv564) [inline plugin/symbols.py]: Prefer to check for keys instead of using exceptions for flow control.

File: tests/test_completion.py [hunk 2]
```diff
@@ -158,37 +168,125 @@ def test_ignores_text_edit(self):
...
+        yield 100  # wait for file to be open
+        handler = CompletionHandler(self.view)
+        handler.last_location = 1
+        handler.last_prefix = ""
+        result = [handler.format_completion(item) for item in clangd_completion_sample]
+        # We should prefer textEdit over insertText. This test covers that.
+        self.assertEqual(
+            result,
+            [
+                ('argc\t  int', 'argc'),  # noqa: E501
```
Comment (tomv564) [inline tests/test_completion.py]: What is needed to get rid of the noqa exceptions?

File: tests/test_completion.py [hunk 3]
```diff
@@ -158,37 +168,125 @@ def test_ignores_text_edit(self):
...
         handler.last_location = 1
         handler.last_prefix = ""
         result = handler.format_completion(item)
-        self.assertEqual(len(result), 2)
-        self.assertEqual("$what", result[0])
-        self.assertEqual("\\$what", result[1])
-        # settings.complete_using_text_edit = False
+        self.assertEqual(result, ('const\t  Keyword', 'const'))
+
+    def test_text_edit1(self):
```
Comment (tomv564) [inline tests/test_completion.py]: Would like to know the test name to describe difference between test_text_edit1 and test_text_edit2

File: plugin/signature_help.py [hunk 4]
```diff
@@ -21,6 +21,22 @@
 from .core.settings import settings
 
 
+def get_documentation(d: 'Dict[str, Any]') -> 'Optional[str]':
```
Comment (tomv564) [inline plugin/signature_help.py]: Unsure if this is the same code as the previous PR, but would be nice to have only 1 instance of this function in this codebase.

Comment (rwols) [reply plugin/signature_help.py]: This is commit c95ae70, cherry-picked from #520. I will address the rest of the review tomorrow.

File: tests/test_completion.py (hunk 2 above)
Comment (rwols) [reply tests/test_completion.py]: I added a handful of newlines where necessary.

File: plugin/signature_help.py (hunk 4 above)
Comment (rwols) [reply plugin/signature_help.py]: This has now been merged and thus is not visible anymore in this PR.

**Pull request merged**